src/daemon/distribution.c

```diff
@@ -0,0 +1,402 @@
+  }
+
+  /* thanks to previous condition we know that if d1 is NULL then d2 is NULL too
+   */
+  if (d1 == NULL) {
+    return true;
+  }
+
+  pthread_mutex_lock(&d1->mutex);
+  pthread_mutex_lock(&d2->mutex);
+
+  if (d1->sum_gauges != d2->sum_gauges) {
+    return false;
+  }
+
+  if (d1->num_buckets != d2->num_buckets) {
+    return false;
+  }
+
+  for (size_t i = 0; i < d1->num_buckets; ++i) {
+    if (d1->buckets[i].max_boundary != d2->buckets[i].max_boundary ||
+        d1->buckets[i].counter != d2->buckets[i].counter) {
+      return false;
+    }
+  }
+
+  pthread_mutex_unlock(&d2->mutex);
+  pthread_mutex_unlock(&d1->mutex);
+
+  return true;
+}
+
+static bucket_t *bucket_new_linear(size_t num_buckets, double size) {
+  bucket_t *buckets = calloc(num_buckets, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  for (size_t i = 0; i < num_buckets - 1; ++i) {
+    buckets[i].max_boundary = (double)(i + 1) * size;
+  }
+
+  buckets[num_buckets - 1].max_boundary = INFINITY;
+
+  return buckets;
+}
+
+static bucket_t *bucket_new_exponential(size_t num_buckets, double base,
+                                        double factor) {
+  bucket_t *buckets = calloc(num_buckets, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  double multiplier = 1.0;
+
+  for (size_t i = 0; i < num_buckets - 1; ++i) {
+    buckets[i].max_boundary = factor * multiplier;
+    multiplier *= base;
+  }
+
+  buckets[num_buckets - 1].max_boundary = INFINITY;
+
+  return buckets;
+}
+
+static bucket_t *bucket_new_custom(size_t num_boundaries,
+                                   const double *custom_buckets_boundaries) {
+  bucket_t *buckets = calloc(num_boundaries + 1, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  if (num_boundaries > 0) {
+    if (custom_buckets_boundaries[0] <= 0 ||
+        custom_buckets_boundaries[0] == INFINITY) {
+      free(buckets);
+      errno = EINVAL;
+      return NULL;
+    }
+
+    buckets[0].max_boundary = custom_buckets_boundaries[0];
+
+    for (size_t i = 1; i < num_boundaries; ++i) {
+      if (custom_buckets_boundaries[i] <= 0 ||
+          custom_buckets_boundaries[i] == INFINITY ||
+          custom_buckets_boundaries[i - 1] >= custom_buckets_boundaries[i]) {
+        free(buckets);
+        errno = EINVAL;
+        return NULL;
+      }
+
+      buckets[i].max_boundary = custom_buckets_boundaries[i];
+    }
+  }
+
+  buckets[num_boundaries].max_boundary = INFINITY;
+
+  return buckets;
+}
+
+distribution_t *distribution_new_linear(size_t num_buckets, double size) {
+  if (num_buckets == 0 || size <= 0) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  distribution_t *d = calloc(1, sizeof(distribution_t));
+
+  if (d == NULL) {
+    return NULL;
+  }
+
+  d->buckets = bucket_new_linear(num_buckets, size);
+
+  if (d->buckets == NULL) {
+    free(d);
+    return NULL;
+  }
+
+  d->num_buckets = num_buckets;
+  pthread_mutex_init(&d->mutex, NULL);
```

Comment:
I initialise the mutex. When you create it, you should initialise it, to be sure that it will work correctly.

https://linux.die.net/man/3/pthread_mutex_init